girder/utility/filesystem_assetstore_adapter.py

```diff
@@ -219,8 +219,9 @@ def finalizeUpload(self, upload, file):
 
         # Store the hash in the upload so that deleting a file won't delete
         # this file
-        upload['sha512'] = hash
-        self.model('upload').update({'_id': upload['_id']}, update={'$set': {'sha512': hash}})
+        if '_id' in upload:
```

Comment:
It seems we are changing some behavior that assetstore implementations may have been relying on -- to be safe, I think we should defer the no-save optimization until 3.0, and save the upload model even in a single-shot case.